docs/tracing.rst

```diff
@@ -0,0 +1,144 @@
+.. _tracing:
+
+Tracing
+=======
+
+.. contents::
+    :depth: 3
+
+
+Overview
+--------
+
+It provides an instrument tracing wrapper function to use to instrument functions manually in pubtools-* projects.
+
+
+Usage
+.....................
+
+Set environment variables
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Following environment variables are used in the module:
+
+- ``OTEL_TRACING``: set ``true`` to enable instrument tracing, otherwise tracing is disabled.
+- ``OTEL_SERVICE_NAME``: required, set the value of the service.name resource attribute. It's
+  expected to be unique within the same namespace.
+
+
+OTEL Exporter
+~~~~~~~~~~~~~~
+
+In order to visualize and analyze telemetry, an exporter is required to export tracing data to
+a backend, e.g: `jaeger <https://www.jaegertracing.io/>`_. As part of OpenTelemetry Python you
+will find many exporters being available. Which exporter should be used depends on usage scenarios, e.g:
+`ConsoleSpanExporter <https://opentelemetry.io/docs/instrumentation/python/exporters/#console-exporter/>`_
+is useful for development and debugging tasks, while
+`OTLPSpanExporter <https://opentelemetry.io/docs/instrumentation/python/exporters/#otlp-endpoint-or-collector/>`_
+can be more suitable on production environment. So choosing different exporters for different scenarios is expected.
+
+In order to have a exporter you expected, the hook :meth:`otel_exporter` is needed to be implemented,
+otherwise `ConsoleSpanExporter <https://opentelemetry.io/docs/instrumentation/python/exporters/#console-exporter/>`_
+will be used.
+
+Instrument tracing for functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Before instrument tracing for functions, :class:`TracingWrapper` should be called once in the service, which
```

Comment:
Could we please remove this requirement?  It seems like an ugly implementation detail.  The user doesn't have to actually do anything with the class except instantiate it, right?

Actually, I'm reading `TracingWrapper` right now and I can't understand why an instance of the class is even needed.  There is never any reference to `TracingWrapper.__instance` anywhere.  Is the point of that code just to do some one-time initialization and there is no technical reason for it to be a class (as opposed to e.g. just a plain function)?

Anyway, I would very much like us to find some way to avoid this and to do this automatically for the user.

Right now, when does the initialization need to happen?  Is it necessary before `@instrument_func` or is it just necessary before one of the instrumented functions is invoked?  I would suggest that instrument_func itself can arrange for this initialization to happen once, when needed, and don't require the caller to care about this.

Reply:
I do see one usage of TracingWrapper in the change, which is that this happens in some tests:

```
tracing_wrapper.processor.force_flush()
```

Would you anticipate that other projects would ever need to do something like that?  If so, that's a reason why it would have to remain in the API.

If it needs to remain in the API because projects would need to use API like the above, here is a different suggestion how to handle it:

There should be a public class named something like `TraceContext` which provides access to the processor (and whatever other otel classes that projects might need to access).

There should be a public function named something like `get_trace_context()` which returns the single global trace context, creating it if necessary; or returning an object which does nothing if tracing is disabled.

The instrument_func decorator, rather than being a standalone function, should be a member of TraceContext.  The way of using it downstream would then be something like:

```
from pubtools.tracing import get_trace_context

ctx = get_trace_context()

class Whatever:
    @ctx.instrument_func(...)
    def some_method_i_want_to_instrument(self):
        ...
```

Note that it forces the caller to obtain a TraceContext before they can instrument anything, so it's not possible to forget to initialize the context.

Reply:
> I do see one usage of TracingWrapper in the change, which is that this happens in some tests:
> 
> ```
> tracing_wrapper.processor.force_flush()
> ```
> 
> Would you anticipate that other projects would ever need to do something like that? If so, that's a reason why it would have to remain in the API.

tracing_wrapper.processor.force_flush() will be used in pubd to make sure all trace data to be sent to exporters before a pub task exits.

Besides that racing_wrapper.provider will be used when enabling RequestsInstrumentor() in pubd.
`
RequestsInstrumentor().instrument(tracer_provider=tracing_wrapper.provider)
 `